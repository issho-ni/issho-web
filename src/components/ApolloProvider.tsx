import { ApolloProvider as _ApolloProvider } from "@apollo/react-common"
import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory"
import ApolloClient from "apollo-client"
import { ApolloLink } from "apollo-link"
import { ErrorHandler, ErrorLink } from "apollo-link-error"
import { HttpLink } from "apollo-link-http"
import * as React from "react"
import { AuthLink } from "../lib/apollo/authLink"
import { SessionContext } from "./SessionProvider"

export interface ApolloProviderProps {
  children: React.ReactNode
}

export interface ApolloProviderState {
  client: ApolloClient<NormalizedCacheObject>
}

export class ApolloProvider extends React.Component<
  ApolloProviderProps,
  ApolloProviderState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>

  constructor(props: ApolloProviderProps) {
    super(props)

    const httpLink = new HttpLink({
      uri: GRAPHQL_ENDPOINT,
    })

    const authLink = new AuthLink(this.getToken)

    const errorLink = new ErrorLink(this.handleError)

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: ApolloLink.from([authLink, errorLink, httpLink]),
    })

    this.state = { client }
  }

  public componentDidUpdate() {
    const { client } = this.state
    client.clearStore()
  }

  public render = () => <_ApolloProvider {...this.props} {...this.state} />

  private getToken = () => this.context.token

  private handleError: ErrorHandler = ({ graphQLErrors = [] }) => {
    for (const err of graphQLErrors) {
      switch (err.message) {
        case "ERR_AUTHENTICATION_REQUIRED":
          this.context.handleLogout()
          break
        default:
          console.error(`[GraphQL Error] ${err.message}`)
      }
    }
  }
}
