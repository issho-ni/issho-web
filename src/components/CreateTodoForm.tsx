import { MutationFunction } from "@apollo/react-common"
import { Form, Formik, FormikActions } from "formik"
import * as React from "react"
import { Field } from "./Field"
import { GET_TODOS } from "./TodoList"

export interface CreateTodoFormProps {
  createTodo: MutationFunction<CreateTodoResult>
}

export interface CreateTodoFormValues {
  text: string
}

export class CreateTodoForm extends React.Component<CreateTodoFormProps> {
  public render() {
    return (
      <Formik<CreateTodoFormValues>
        initialValues={{ text: "" }}
        onSubmit={this.handleCreateTodo}
      >
        {() => (
          <Form>
            <Field type="text" name="text" required />
            <button type="submit">Create Todo</button>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateTodo = (
    variables: CreateTodoFormValues,
    actions: FormikActions<CreateTodoFormValues>
  ) =>
    this.props
      .createTodo({
        optimisticResponse: {
          __typename: "Mutation",
          createTodo: {
            __typename: "Todo",
            completedAt: null,
            createdAt: new Date(),
            id: "temp",
            ...variables,
          },
        },
        update: (proxy, { data: { createTodo } }) => {
          const data = proxy.readQuery<GetTodosResult>({ query: GET_TODOS })
          data.getTodos.push(createTodo)
          proxy.writeQuery({ query: GET_TODOS, data })
        },
        variables,
      })
      .then(() => actions.resetForm({ text: "" }))
}
