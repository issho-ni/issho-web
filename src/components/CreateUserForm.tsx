import { Field, Form, Formik } from "formik"
import * as React from "react"
import { MutationFn } from "react-apollo"
import { Link } from "react-router-dom"
import { clearState } from "../utils/state"
import { SessionContext } from "./SessionContext"

export interface CreateUserFormProps {
  createUser: MutationFn<LoginUserResult>
}

export interface CreateUserFormState {
  error?: string
}

export interface CreateUserFormValues {
  email: string
  name: string
  password: string
}

export class CreateUserForm extends React.Component<
  CreateUserFormProps,
  CreateUserFormState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>
  public state: CreateUserFormState = {}

  public render() {
    const { error } = this.state

    return (
      <Formik<CreateUserFormValues>
        initialValues={{ email: "", name: "", password: "" }}
        onSubmit={this.handleCreateUser}
      >
        {() => (
          <Form>
            {error ? <span>{error}</span> : ""}
            <Field type="text" name="name" autoComplete="name" required />
            <Field type="email" name="email" autoComplete="email" required />
            <Field
              type="password"
              name="password"
              autoComplete="new-password"
              required
            />
            <button type="submit">Create Account</button>
            <Link to="/login">Log In</Link>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateUser = (variables: CreateUserFormValues) =>
    this.props
      .createUser({ variables })
      .then(clearState(this, ["error"]))
      .then(this.context.handleLogin)
      .catch(() =>
        this.setState({
          error: "Could not create an account with that e-mail address",
        })
      )
}
