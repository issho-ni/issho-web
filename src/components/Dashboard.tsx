import * as React from "react"
import { Logout } from "./Logout"

export class Dashboard extends React.Component {
  public render() {
    return (
      <div>
        <h2>Dashboard</h2>
        <Logout />
      </div>
    )
  }
}
