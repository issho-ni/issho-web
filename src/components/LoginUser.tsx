import { Field, Form, Formik } from "formik"
import gql from "graphql-tag"
import * as React from "react"
import { Mutation, MutationFn } from "react-apollo"
import { SessionContext, SessionLoginHandler } from "./SessionContext"

const LOGIN_USER = gql`
  mutation LoginUser($email: String!, $password: String!) {
    loginUser(input: { email: $email, password: $password }) {
      token
      user {
        email
        name
      }
    }
  }
`

interface LoginUserState {
  error?: string
}

interface LoginUserValues {
  email: string
  password: string
}

export class LoginUser extends React.Component<{}, LoginUserState> {
  constructor(props) {
    super(props)
    this.state = {}
  }

  public render() {
    const { error } = this.state

    return (
      <SessionContext.Consumer>
        {({ handleLogin }) => (
          <Mutation mutation={LOGIN_USER}>
            {loginUser => (
              <Formik<LoginUserValues>
                initialValues={{ email: "", password: "" }}
                onSubmit={this.handleLogin(loginUser, handleLogin)}
              >
                {() => (
                  <Form>
                    {error ? <span>{error}</span> : ""}
                    <Field
                      type="email"
                      name="email"
                      autoComplete="email"
                      required
                    />
                    <Field
                      type="password"
                      name="password"
                      autoComplete="current-password"
                      required
                    />
                    <button type="submit">Log In</button>
                  </Form>
                )}
              </Formik>
            )}
          </Mutation>
        )}
      </SessionContext.Consumer>
    )
  }

  private handleLogin = (
    loginUser: MutationFn,
    handleLogin: SessionLoginHandler
  ) => (values: LoginUserValues) =>
    loginUser({ variables: values })
      .then(data => {
        this.setState({ error: null })
        return data
      })
      .then(handleLogin)
      .catch(() => this.setState({ error: "Incorrect username or password" }))
}
