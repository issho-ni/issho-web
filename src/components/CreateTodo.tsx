import gql from "graphql-tag"
import * as React from "react"
import { Mutation } from "react-apollo"

const CreateTodoForm = React.lazy(() =>
  import("./CreateTodoForm").then(module => ({
    default: module.CreateTodoForm,
  }))
)

export const CREATE_TODO = gql`
  mutation CreateTodo($text: String!) {
    createTodo(input: { text: $text }) {
      id
      text
      createdAt
      completedAt
    }
  }
`

export class CreateTodo extends React.Component {
  public render() {
    return (
      <Mutation mutation={CREATE_TODO}>
        {createTodo => <CreateTodoForm {...{ createTodo }} />}
      </Mutation>
    )
  }
}
