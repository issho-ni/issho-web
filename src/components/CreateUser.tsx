import gql from "graphql-tag"
import * as React from "react"
import { Mutation } from "react-apollo"

const CreateUserForm = React.lazy(() =>
  import("./CreateUserForm").then(module => ({
    default: module.CreateUserForm,
  }))
)

export const CREATE_USER = gql`
  mutation CreateUser($email: String!, $name: String!, $password: String!) {
    createUser(input: { email: $email, name: $name, password: $password }) {
      token
      user {
        email
        name
      }
    }
  }
`

export class CreateUser extends React.Component {
  public render() {
    return (
      <Mutation<CreateUserResult> mutation={CREATE_USER}>
        {createUser => <CreateUserForm {...{ createUser }} />}
      </Mutation>
    )
  }
}
