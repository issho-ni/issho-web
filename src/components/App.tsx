import * as React from "react"
import { BrowserRouter } from "react-router-dom"

const ApolloProvider = React.lazy(() =>
  import("./ApolloProvider").then(module => ({
    default: module.ApolloProvider,
  }))
)
const CreateUser = React.lazy(() =>
  import("./CreateUser").then(module => ({ default: module.CreateUser }))
)
const Dashboard = React.lazy(() =>
  import("./Dashboard").then(module => ({ default: module.Dashboard }))
)
const LoginUser = React.lazy(() =>
  import("./LoginUser").then(module => ({ default: module.LoginUser }))
)
const ProtectedRoute = React.lazy(() =>
  import("./ProtectedRoute").then(module => ({
    default: module.ProtectedRoute,
  }))
)
const SessionProvider = React.lazy(() =>
  import("./SessionProvider").then(module => ({
    default: module.SessionProvider,
  }))
)

export class App extends React.Component {
  public render() {
    return (
      <React.Suspense fallback={<h1>Loadingâ€¦</h1>}>
        <SessionProvider>
          <ApolloProvider>
            <BrowserRouter>
              <ProtectedRoute
                path="/"
                authenticated
                exact
                component={Dashboard}
              />
              <ProtectedRoute path="/join" component={CreateUser} />
              <ProtectedRoute path="/login" component={LoginUser} />
            </BrowserRouter>
          </ApolloProvider>
        </SessionProvider>
      </React.Suspense>
    )
  }
}
