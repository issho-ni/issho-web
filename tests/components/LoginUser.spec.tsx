import * as React from "react"
import { fireEvent, wait, waitForElement } from "react-testing-library"
import { LOGIN_USER, LoginUser } from "../../src/components/LoginUser"
import { render } from "../TestWrapper"

const mocks = [
  {
    request: {
      query: LOGIN_USER,
      variables: {
        email: "test@test",
        password: "P@55w0rd",
      },
    },
    result: {
      data: {
        loginUser: {
          token: "not-a-valid-jwt",
          user: {
            email: "test@test",
            id: "5265b472-4b64-7014-9242-3faa40163fde",
            name: "Test",
          },
        },
      },
    },
  },
]

const component = () => render(<LoginUser />, { mocks })

describe("LoginUser", () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it("renders without error", async () => {
    const { container, getByText } = component()
    await wait(() => getByText(/log in/i, { selector: "button" }))

    expect(container).toMatchSnapshot()
  })

  it("submits the entered values", async () => {
    const { email, password } = mocks[0].request.variables
    const { getByLabelText, getByText } = component()

    const emailInput = await waitForElement(() => getByLabelText(/email/i))
    fireEvent.change(emailInput, {
      target: { name: "email", value: email },
    })

    const passwordInput = await waitForElement(() =>
      getByLabelText(/password/i)
    )
    fireEvent.change(passwordInput, {
      target: { name: "password", value: password },
    })

    const button = await waitForElement(() =>
      getByText(/log in/i, { selector: "button" })
    )
    fireEvent.click(button)

    await wait(() =>
      expect(localStorage.setItem).toBeCalledWith(
        "session",
        JSON.stringify(mocks[0].result.data.loginUser)
      )
    )
  })

  it("displays errors", async () => {
    const { getByText } = component()

    const button = await waitForElement(() =>
      getByText(/log in/i, { selector: "button" })
    )
    fireEvent.click(button)

    await wait(() => getByText("Incorrect e-mail address or password"))
    expect(localStorage.setItem).not.toBeCalled()
  })
})
