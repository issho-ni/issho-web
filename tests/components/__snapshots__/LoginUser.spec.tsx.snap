// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginUser renders without error 1`] = `
<div>
  <form>
    
    <div>
      <label
        for="email"
      >
        email
      </label>
      <input
        autocomplete="email"
        id="email"
        name="email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
        password
      </label>
      <input
        autocomplete="current-password"
        id="password"
        name="password"
        required=""
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Log In
    </button>
    <a
      href="/join"
    >
      Create Account
    </a>
  </form>
</div>
`;
